### PR TITLE
Add process segment definition iterator

### PR DESCRIPTION
**Pull request by Levi-Armstrong**

### CODE REVIEW

**Comment by Levi-Armstrong** [comment]
@mpowelson Here is a start to adding the iterator to some of the process planners classes. I wanted to get this out sooner because it will just be a copy and past to the rest of the classes with some minor name changes. Let me know if you see any issues or change I should make.

**Comment by Levi-Armstrong** [comment]
I will also go through and document add test to verify that everything works as expected.

File: tesseract/tesseract_planning/tesseract_process_planners/include/tesseract_process_planners/process_segment_definition.h [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+  }
+
+  bool isProcess(const std::size_t index) const
+  {
+    return ((pos_ >= approach.size()) && (pos_ < (approach.size() + process.size())))
+  }
+
+  bool isDeparture(const std::size_t index) const
+  {
+    return ((pos_ >= (approach.size() + process.size())) && (pos_ < (approach.size() + process.size() + departure.size())))
```
[the remark this answers is not in the document]
Comment (Levi-Armstrong) [reply tesseract/tesseract_planning/tesseract_process_planners/include/tesseract_process_planners/process_segment_definition.h]: fixed

**Comment by Levi-Armstrong** [comment]
@mpowelson I have add documentation and believe to have finished the iterator class to support random access. I have also added unit test for the custom iterator.

**Comment by Levi-Armstrong** [comment]
@gavanderhoorn I have never implemented a custom iterator, do you mind taking a look?

File: tesseract/tesseract_planning/tesseract_process_planners/test/custom_iterator_unit.cpp [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  psd.erase(psd.begin(), psd.begin() + 3);
+  check.erase(check.begin(), check.begin() + 3);
+  EXPECT_TRUE(psd.size() == 11);
+  EXPECT_TRUE(psd.approach.size() == 1);
+
+  for (size_t i = 0; i < check.size(); ++i)
+  {
+    EXPECT_TRUE(check[i] == psd[i]);
```
Comment (mpowelson) [inline tesseract/tesseract_planning/tesseract_process_planners/test/custom_iterator_unit.cpp]: Should .at(i) be tested as well?

File: tesseract/tesseract_planning/tesseract_process_planners/test/custom_iterator_unit.cpp [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+  long d = std::distance(psd.begin(), f1);
+  EXPECT_TRUE(d == 4);
+
+  EXPECT_TRUE(psd.size() == 15);
+  EXPECT_TRUE(psd.isApproach(4));
+  EXPECT_FALSE(psd.isApproach(5));
+  EXPECT_TRUE(psd.isProcess(9));
+  EXPECT_FALSE(psd.isProcess(10));
+  EXPECT_TRUE(psd.isDeparture(10));
```
[the remark this answers is not in the document]
Comment (Levi-Armstrong) [reply tesseract/tesseract_planning/tesseract_process_planners/test/custom_iterator_unit.cpp]: I will add the additional tests

File: tesseract/tesseract_planning/tesseract_process_planners/test/custom_iterator_unit.cpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  psd.erase(psd.begin(), psd.begin() + 3);
+  check.erase(check.begin(), check.begin() + 3);
+  EXPECT_TRUE(psd.size() == 11);
+  EXPECT_TRUE(psd.approach.size() == 1);
+
+  for (size_t i = 0; i < check.size(); ++i)
+  {
+    EXPECT_TRUE(check[i] == psd[i]);
+  }
+
```
[the remark this answers is not in the document]
Comment (Levi-Armstrong) [reply tesseract/tesseract_planning/tesseract_process_planners/test/custom_iterator_unit.cpp]: I will add a few test that use end().

File: tesseract/tesseract_planning/tesseract_process_planners/test/custom_iterator_unit.cpp (hunk 2 above)
Comment (Levi-Armstrong) [reply tesseract/tesseract_planning/tesseract_process_planners/test/custom_iterator_unit.cpp]: I know that std::vector has `.at` method but it did no seem to be required and I could not find much documentation on custom iterators. The `.at` method allow access to elements in a const function where `[]` does not for std::vector. The iterator I have implemented has two `operator[]` implementation one that is const and one that is not which I believe is why the `.at` is not required. I will create a test class with a const function and make sure I can access elements.

Comment (mpowelson) [reply tesseract/tesseract_planning/tesseract_process_planners/test/custom_iterator_unit.cpp]: So I looked at it, and I came to the conclusion that it was not part of an [iterator](http://www.cplusplus.com/reference/iterator/). However, then I thought I saw where you implemented it. Looking back, I think I was looking at operator*.

File: tesseract/tesseract_planning/tesseract_process_planners/test/custom_iterator_unit.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Levi-Armstrong) [reply tesseract/tesseract_planning/tesseract_process_planners/test/custom_iterator_unit.cpp]: Updated

File: tesseract/tesseract_planning/tesseract_process_planners/src/tesseract_process_planners/process_segment_definition.cpp [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+  return ret;
+}
+
+
+ProcessSegmentDefinition::Iterator::Iterator(ProcessSegmentDefinition& container, std::size_t pos) : pos_(pos), container_(&container)
+{
+
+}
+
+// this method must be defined after the definition of IntVector
```
[the remark this answers is not in the document]
Comment (Levi-Armstrong) [reply tesseract/tesseract_planning/tesseract_process_planners/src/tesseract_process_planners/process_segment_definition.cpp]: fixed

File: tesseract/tesseract_planning/tesseract_process_planners/test/custom_iterator_unit.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Levi-Armstrong) [reply tesseract/tesseract_planning/tesseract_process_planners/test/custom_iterator_unit.cpp]: Test added

File: tesseract/tesseract_planning/tesseract_process_planners/test/custom_iterator_unit.cpp (hunk 2 above)
Comment (Levi-Armstrong) [reply tesseract/tesseract_planning/tesseract_process_planners/test/custom_iterator_unit.cpp]: I added the at method to the class.

**Comment by Levi-Armstrong** [comment]
@mpowelson I believe I have addressed the requested changes.

File: tesseract/tesseract_planning/tesseract_process_planners/test/custom_iterator_unit.cpp (hunk 2 above)
Comment (Levi-Armstrong) [reply tesseract/tesseract_planning/tesseract_process_planners/test/custom_iterator_unit.cpp]: I will add the test

Comment (Levi-Armstrong) [reply tesseract/tesseract_planning/tesseract_process_planners/test/custom_iterator_unit.cpp]: The tests have been added for the `at` method.

**Comment by mpowelson** [comment]
I say go for it.

Also, a collective thank you from all of the people that have manually iterated over a processDefinition.

**Pull request merged**